src/runtime_src/core/common/sensor.cpp

```diff
@@ -211,15 +215,96 @@ read_electrical(const xrt_core::device * device)
+    auto power  = xrt_core::device_query<xq::sdm_sensor_info>(device, (int)POWER_REQ);
+    if (current.empty() ^ voltage.empty() ^ power.empty()) {
+      sensor_array.push_back(std::make_pair("", pt.put("data_driven", false)));
+      is_data_driven = false;
+    }
+    else {
+      pt.put("data_driven", true);
+      std::string val;
+      for(size_t i = 0; i < current.size(); i++)
+      {
+        auto tmp = current[i];
+        pt.put("label", tmp.label);
+        pt.put("instantaneous", xrt_core::utils::format_base10_shiftdown3(tmp.input));
+        pt.put("max", xrt_core::utils::format_base10_shiftdown3(tmp.max));
+        pt.put("average", xrt_core::utils::format_base10_shiftdown3(tmp.average));
+        pt.put("highest", xrt_core::utils::format_base10_shiftdown3(tmp.highest));
+        pt.put("current.is_present", true);
+        pt.put("voltage.is_present", false);
+        pt.put("power.is_present", false);
+        sensor_array.push_back(std::make_pair("", pt));
+      }
+      for(size_t i = 0; i < voltage.size(); i++)
+      {
+        auto tmp = voltage[i];
+        pt.put("label", tmp.label);
+        pt.put("instantaneous", xrt_core::utils::format_base10_shiftdown3(tmp.input));
+        pt.put("max", xrt_core::utils::format_base10_shiftdown3(tmp.max));
+        pt.put("average", xrt_core::utils::format_base10_shiftdown3(tmp.average));
+        pt.put("highest", xrt_core::utils::format_base10_shiftdown3(tmp.highest));
+        pt.put("voltage.is_present", true);
+        pt.put("current.is_present", false);
+        pt.put("power.is_present", false);
+        sensor_array.push_back(std::make_pair("", pt));
+      }
+      for(size_t i = 0; i < power.size(); i++)
+      {
+        auto tmp = power[i];
+        pt.put("label", tmp.label);
+        pt.put("instantaneous", xrt_core::utils::format_base10_shiftdown6(tmp.input));
+        pt.put("max", xrt_core::utils::format_base10_shiftdown6(tmp.max));
+        pt.put("average", xrt_core::utils::format_base10_shiftdown6(tmp.average));
+        pt.put("highest", xrt_core::utils::format_base10_shiftdown6(tmp.highest));
+        pt.put("power.is_present", true);
+        pt.put("voltage.is_present", false);
+        pt.put("current.is_present", false);
+        sensor_array.push_back(std::make_pair("", pt));
+        if (!strcmp((tmp.label).c_str(), "POWER"))
+          val = xrt_core::utils::format_base10_shiftdown6(tmp.input);
+      }
+      root.add_child("power_rails", sensor_array);
+      root.put("power_consumption_watts", val);
+      root.put("power_consumption_max_watts", "NA");
+      root.put("power_consumption_warning", "NA");
+    }
+  }
+  catch(const xrt_core::query::no_such_key&) {
+    is_data_driven = false;
+  }
+  catch (const std::exception& ex) {
+    sensor_array.push_back(std::make_pair("", pt.put("error_msg", ex.what())));
+  }
+
+  if (!is_data_driven)
+    return read_legacy_electrical(device);
```

Comment:
How about just return this from the first catch?

Reply:
it may miss case where ew sdr model returns empty output then code try to access legacy